src/index/compat.rs

```diff
@@ -0,0 +1,128 @@
+                .parse::<u32>()
+                .unwrap();
+            let ef_construction = original
+                .get("ef_construction")
+                .unwrap_or(&String::from("64"))
+                .parse::<usize>()
+                .unwrap();
+            config.indexing = IndexingOptions::Hnsw(HnswIndexingOptions {
+                m,
+                ef_construction,
+                ..Default::default()
+            });
+        }
+        Ok("ivfflat") => {
+            let nlist = original
+                .get("list")
+                .unwrap_or(&String::from("100"))
+                .parse::<u32>()
+                .unwrap();
+            config.indexing = IndexingOptions::Ivf(IvfIndexingOptions {
+                nlist,
+                ..Default::default()
+            });
+        }
+        Ok(other) => SessionError::BadLiteral {
+            hint: format!("IndexType/accessMethod {other} not allowed").to_string(),
+        }
+        .friendly(),
+        Err(e) => SessionError::BadLiteral {
+            hint: format!("Bad IndexType/accessMethod: {e}").to_string(),
+        }
+        .friendly(),
+    }
+    let toml = toml::to_string(&config).unwrap_or_else(|_| {
+        SessionError::Internal {
+            hint: "Failed to serialize toml config".to_string(),
+        }
+        .friendly()
+    });
+    let arg = pgrx::pg_sys::makeString(toml.as_pg_cstr());
+    let elem = pgrx::pg_sys::makeDefElem("options".as_pg_cstr(), arg as _, -1);
+    (*istmt).options = pack_element_list(elem);
+    (*istmt).accessMethod = "vectors".as_pg_cstr();
+}
+
+unsafe fn rewrite_opclass(istmt: *mut pgrx::pg_sys::IndexStmt) {
+    let params = (*istmt).indexParams;
+    let elem: *mut pgrx::pg_sys::IndexElem = *downcast_list::<*mut c_void>(params)
+        .get(0)
+        .unwrap_or_else(|| {
+            SessionError::Internal {
+                hint: "failed to get element config".to_string(),
+            }
+            .friendly()
+        }) as _;
+
+    let opclass_name = *downcast_list::<*mut c_void>((*elem).opclass)
+        .get(0)
+        .unwrap_or_else(|| {
+            SessionError::Internal {
+                hint: "failed to get opclass".to_string(),
+            }
+            .friendly()
+        });
+    #[cfg(any(feature = "pg15", feature = "pg16"))]
+    let opclass_ptr = (*(opclass_name as *mut pgrx::pg_sys::String)).sval;
+    #[cfg(any(feature = "pg12", feature = "pg13", feature = "pg14"))]
+    let opclass_ptr = (*(opclass_name as *mut pgrx::pg_sys::Value)).val.str_;
+    let opclass = match CStr::from_ptr(opclass_ptr).to_str() {
+        Ok("vector_l2_ops") => "vector_l2_ops",
+        Ok("vector_ip_ops") => "vector_dot_ops",
+        Ok("vector_cosine_ops") => "vector_cos_ops",
+        Ok(other) => SessionError::BadLiteral {
+            hint: format!("Operator {other} not allowed").to_string(),
+        }
+        .friendly(),
+        Err(e) => SessionError::BadLiteral {
```

Comment:
Keep the orginal value, do not panic.

Reply:
Also you should see usage of `BadLiteral`. Errors are semantic, not just a combination of words.